In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
#Importing data_products
df_prods= pd.read_csv(r'C:\Users\ahamr\Desktop\12-2023 Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

## Mixed-Type Data

In [5]:
#Creating a test df
df_test = pd.DataFrame()

In [6]:
#Creating a mix-type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mix types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\ahamr\AppData\Local\Temp\ipykernel_3420\2922229538.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [9]:
#Changing data type
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [10]:
#Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Creating a subset with missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
#Addressing Missing Values
df_prods.shape

(49693, 5)

In [14]:
#Creating a clean df
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

## Duplicates

In [16]:
#Looking for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
#Creating a df with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
# Exporting the df_prods
path=r'C:\Users\ahamr\Desktop\12-2023 Instacart Basket Analysis'
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## Task 4.5

In [22]:
# Importing prepared data
df_ords = pd.read_csv(r'C:\Users\ahamr\Desktop\12-2023 Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [23]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour  days_since_prior_order  
0           8                     NaN  
1           7                    15.0  
2          12                    21.0  
3           7                    29.0  
4          15                    28.0

In [24]:
# Run the df.describe() function on your df_ords.
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_hour  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

Investigating our frequency table, in particular the max and min section, it appears that days_since_prior_order has strange results. We can see that the min- 0 and max- 3, however, the 50% is 7, which is higher than the max.

In [25]:
# 3. Checking for mixed-type data

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
   print (col)

C:\Users\ahamr\AppData\Local\Temp\ipykernel_3420\3180090806.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\ahamr\AppData\Local\Temp\ipykernel_3420\3180090806.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\ahamr\AppData\Local\Temp\ipykernel_3420\3180090806.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\ahamr\AppData\Local\Temp\ipykernel_3420\3180090806.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\ahamr\AppData\Local\Tem

In [26]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour                  int64
days_since_prior_order    float64
dtype: object

There are no mixed-type data in df_ords.

In [27]:
# 5.Checking for Missing Values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour                     0
days_since_prior_order    206209
dtype: int64

In [28]:
# Creating a subset with missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [29]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour  days_since_prior_order  
0                         2           8                     NaN  
11                        2          11                     NaN  
26                        1          14                     NaN  
39                        6          11                     NaN  
45                        3          12                     NaN  
...                     ...         ...                     ...  
3420930                   4          12                     NaN  
3420934                   3          18                     NaN  
3421002                   6          19                     NaN  
3421019                   1          15                     NaN  
3421069                   3          11                     NaN  

[206209 rows x 8 columns]

We have 206209 rows of missing values of the days_since_prior_order column. This is not a significant number considering all records of this column. Also, we can see that the missing values have the same order_number - 1. So it can be that there is no record because the customers were making their first orders. In my opinion, we can leave it empty or impute the mode value.

In [30]:
# 6. Address the missing values using an appropriate method.
df_ords['order_number'].value_counts(dropna = False) 

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

As we see, the number of first orders is the same as the missing records. So I consider leaving this column with missing values.

In [31]:
# 7. Checking for duplicate values in df_ords
df_dups = df_ords[df_ords.duplicated()]

In [32]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour, days_since_prior_order]
Index: []

No duplicates found in df_ords data frame.

In [33]:
# Exporting df_ords
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [34]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_hour  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

In [35]:
df_ords.describe().astype(int)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
count     3421083   3421083  3421083       3421083             3421083   
mean      1710541   1710542   102978            17                   2   
std        987581    987581    59533            17                   2   
min             0         1        1             1                   0   
25%        855270    855271    51394             5                   1   
50%       1710541   1710542   102689            11                   3   
75%       2565811   2565812   154385            23                   5   
max       3421082   3421083   206209           100                   6   

       order_hour  days_since_prior_order  
count     3421083                 3214874  
mean           13                      11  
std             4                       9  
min             0                       0  
25%            10                       4  
50%            13                       7  
75%            16                      15  
max            23                      30